net_connection/src/protocol_wrapper.rs

```diff
@@ -0,0 +1,433 @@
+//! This module provides a higher level interface to p2p / network messaging
+//! basically handles serialization / deserialization from / to the core
+//! protocol message types (NamedBinary and Json).
+
+use serde_json;
+
+use super::protocol::Protocol;
+
+/// High level p2p / network message
+#[derive(Debug, Clone, PartialEq)]
+pub enum ProtocolWrapper {
+    /// if we don't have a high-level repr, just store the base protocol
+    RawProtocol(Protocol),
+
+    /// [send] request the current state from the p2p module
+    RequestState,
+
+    /// [recv] p2p module is telling us the current state
+    State(StateData),
+
+    /// [send] request the default config from the p2p module
+    RequestDefaultConfig,
+
+    /// [recv] the default config from the p2p module
+    DefaultConfig(ConfigData),
+
+    /// [send] set the p2p config
+    SetConfig(ConfigData),
+
+    /// [send] connect to the specified multiaddr
+    Connect(ConnectData),
+
+    /// [recv] notification of a peer connected
+    PeerConnected(PeerConnectData),
+
+    /// [send] send a message to another node on the network
+    SendMessage(SendData),
+
+    /// [recv] recv the response back from a previous `SendMessage`
+    SendResult(SendResultData),
+
+    /// [recv] another node has sent us a message
+    HandleSend(HandleSendData),
+
+    /// [send] send our response to a previous `HandleSend`
+    HandleSendResult(SendResultData),
+}
+
+/// upgrade a Protocol reference to a ProtocolWrapper instance
+impl<'a> From<&'a Protocol> for ProtocolWrapper {
+    fn from(p: &'a Protocol) -> Self {
+        if let Protocol::Json(json) = p {
+            let json: serde_json::Value = serde_json::from_str(json.into()).unwrap();
```

Comment:
hmm, this doesn't feel right

firstly an unwrap is dangerous when deserializing JSON as it quite commonly does not deserialize, especially as our data structures are unstable and tend to drift from month to month atm

secondly, could this be represented with an enum that we do `MyEnum::try_from(json)?` against
the matching can then be done against the enum variants rather than raw serde traversals

Reply:
:+1: updated to try_from